Source/WebGPU/WebGPU/RenderBundleEncoder.mm

```diff
@@ -67,51 +67,87 @@
+            auto* pvertexOffsets = pipelineLayout.vertexOffsets(bindGroupIndex, kvp.value);
+            if (pvertexOffsets && pvertexOffsets->size()) {
+                auto& vertexOffsets = *pvertexOffsets;
+                auto startIndex = sizeof(uint32_t) * pipelineLayout.vertexOffsetForBindGroup(bindGroupIndex);
+                auto bytesToCopy = sizeof(vertexOffsets[0]) * vertexOffsets.size();
+                RELEASE_ASSERT(bytesToCopy <= maxBufferLength - (startIndex + bufferOffset));
+                memcpy(&vertexBufferContents[startIndex], &vertexOffsets[0], bytesToCopy);
+            }
         }
 
-        auto* pfragmentOffsets = pipelineLayout.fragmentOffsets(bindGroupIndex, kvp.value);
-        if (pfragmentOffsets && pfragmentOffsets->size()) {
-            auto& fragmentOffsets = *pfragmentOffsets;
-            auto startIndex = pipelineLayout.fragmentOffsetForBindGroup(bindGroupIndex);
-            RELEASE_ASSERT(fragmentOffsets.size() <= m_fragmentDynamicOffsets.size() + startIndex);
-            memcpy(&m_fragmentDynamicOffsets[startIndex], &fragmentOffsets[0], sizeof(fragmentOffsets[0]) * fragmentOffsets.size());
+        if (m_dynamicOffsetsFragmentBuffer) {
+            auto maxBufferLength = m_dynamicOffsetsVertexBuffer.length;
+            auto bufferOffset = fragmentDynamicOffset;
+            uint8_t* fragmentBufferContents = static_cast<uint8_t*>(m_dynamicOffsetsFragmentBuffer.contents) + bufferOffset;
+            auto* pfragmentOffsets = pipelineLayout.fragmentOffsets(bindGroupIndex, kvp.value);
+            if (pfragmentOffsets && pfragmentOffsets->size()) {
+                auto& fragmentOffsets = *pfragmentOffsets;
+                auto startIndex = sizeof(uint32_t) * pipelineLayout.fragmentOffsetForBindGroup(bindGroupIndex);
+                auto bytesToCopy = sizeof(fragmentOffsets[0]) * fragmentOffsets.size();
+                RELEASE_ASSERT(bytesToCopy <= maxBufferLength - (startIndex + bufferOffset));
+                memcpy(&fragmentBufferContents[startIndex], &fragmentOffsets[0], bytesToCopy);
+            }
         }
     }
 
-    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=262208 implement dynamic offsets in render bundles
+    if (m_dynamicOffsetsVertexBuffer)
+        [icbCommand setVertexBuffer:m_dynamicOffsetsVertexBuffer offset:vertexDynamicOffset atIndex:m_device->maxBuffersPlusVertexBuffersForVertexStage()];
+
+    if (m_dynamicOffsetsFragmentBuffer)
+        [icbCommand setFragmentBuffer:m_dynamicOffsetsFragmentBuffer offset:fragmentDynamicOffset atIndex:m_device->maxBuffersForFragmentStage()];
 
     m_bindGroupDynamicOffsets.clear();
 }
 
 void RenderBundleEncoder::draw(uint32_t vertexCount, uint32_t instanceCount, uint32_t firstVertex, uint32_t firstInstance)
 {
-    if (id<MTLIndirectRenderCommand> icbCommand = currentRenderCommand()) {
-        executePreDrawCommands();
+    executePreDrawCommands();
```

Comment:
Does this need to be computed for every draw, or can the calculations be cached?

Reply:
I think it needs to happen before every draw because the vertex buffers can change between draws.

And Metal ICBs appear to require all buffers in use to call setVertexBuffer: for each command, whereas WebGPU maintains a persistent state across commands

Reply:
which is best illustrated through https://webgpu.github.io/webgpu-samples/samples/renderBundles

```
    passEncoder.setPipeline(pipeline);
    passEncoder.setBindGroup(0, frameBindGroup);

    // Loop through every renderable object and draw them individually.
    // (Because many of these meshes are repeated, with only the transforms
    // differing, instancing would be highly effective here. This sample
    // intentionally avoids using instancing in order to emulate a more complex
    // scene, which helps demonstrate the potential time savings a render bundle
    // can provide.)
    let count = 0;
    for (const renderable of renderables) {
      passEncoder.setBindGroup(1, renderable.bindGroup);
      passEncoder.setVertexBuffer(0, renderable.vertices);
      passEncoder.setIndexBuffer(renderable.indices, 'uint16');
      passEncoder.drawIndexed(renderable.indexCount);

      if (++count > settings.asteroidCount) {
        break;
      }
    }
```
`setBindGroup(0` is only called once and the expectation is that bind group is persistent throughout the draw calls